docs/proposals/26-workload-identity.md

```diff
@@ -0,0 +1,514 @@
+    type: aws # Required field.
+    config: # Optional field of type []byte, extensions can make it mandatory via admission webhooks.
+      apiVersion: aws.security.gardener.cloud/v1alpha1
+      kind: Config
+      iamRoleARN: arn:aws:iam::112233445566:role/gardener-dev
+status:
+  sub: gardener.cloud:workloadidentity:garden-local:banana-testing:12b580fe-1f74-4195-852b-e1a74b03496a
+```
+
+JWTs will be available when the clients send `create` requests on the
+`WorkloadIdentity/token` subresource. As the clients will be providing various
+custom information that will be used for the generation of the JWT, yet another
+custom resource `TokenRequest` in the API group `security.gardener.cloud` will
+be used. It is envisioned this resource to contain just metadata for the context
+where the JWT is being used, e.g. shoot or backup entry identifier. Gardener API
+server must verify the provided metadata and it can enhance the JWT with
+additional information derived from the context, for example with information
+for the project and the seed of the shoot cluster. Gardener API can also add
+global information like a landscape identifier. As `WorkloadIdentity` already
+specifies duration of the token, `TokenRequest` will not feature such field.
+`TokenRequest` resources will never be persisted in the storage layer, the
+generated token will be written in the `.status.token` field and returned to the
+client as response. The expiration timestamp of the token will be also available
+in the status via the `.status.expirationTimestamp` field.
+
+```yaml
+apiVersion: security.gardener.cloud/v1alpha1
+kind: TokenRequest
+metadata:
+  name: ""
+  namespace: ""
+spec:
+  contextObject: # Optional field, various metadata about context of use of the token
+    apiVersion: core.gardener.cloud/v1beta1
+    kind: Shoot
+    name: foo
+    namespace: garden-local
+    uid: 54d09554-6a68-4f46-a23a-e3592385d820
+status:
+  token: eyJhbGciOiJ....OkBBrVWA # The generated OIDC token
+  expirationTimestamp: 2024-02-09T16:35:02Z
+```
+
+#### Shoot API
+
+Shoot API will not undergo major changes, it will keep referring to
+SecretBindings in the same namespace by their name, however the `SecretBinding`
+API will be extended in the following way to enable usage of `WorkloadIdentity`
+as infrastructure service credentials:
+
+- `SecretBinding.secretRef` field will be made optional and mutable.
+- new optional field `SecretBinding.workloadIdentityRef` will be introduced, it
+  will refer to a `WorkloadIdentity` resource by its name and namespace. The
+  value of the field will not be immutable and will allow existing shoots to
+  change the workloadidentity they are using to allow rotation by the Gardener
+  users.
+- validation will ensure that either `secretRef` or `workloadIdentityRef` is
+  set, but not both.
+- on update of the `workloadIdentityRef`, a validation will ensure that both the
+  old and the new `WorkloadIdentity` are for the same cloud provider account.
```

Comment:
How can this be validated from GAPI perspective?

Reply:
Thanks for the explanation. This brings me back to @rfranzke's question and how GAPI can validate anything specific in `WorkloadIdentity.spec.targetSystem.config` since this is usually a blackbox for `gardener/gardener`? Do the admission controllers of extensions need to take over here?

Reply:
Yes, by `a validation will ensure` I did not mean just static validations executed by the Gardener API, but also dynamic validations handled by extension webhooks. I will try to re-word and reflect this into the GEP.

Reply:
Done with 937fda954b3884e8f2a8c7e1795726942388e4b7